### PR TITLE
KAFKA-17413; Re-introduce `group.version` feature flag

### PR DESCRIPTION
**Pull request by dajac**
This patch re-introduces the `group.version` feature flag and gates the new consumer rebalance protocol with it. The `group.version` feature flag is attached to the metadata version `4.0-IV0` and it is marked as production ready. This allows system tests to pick it up directly by default without requiring to set `unstable.feature.versions.enable` in all of them. This is fine because we don't plan to do any incompatible changes before 4.0.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (dajac) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:212]: I reordered the MVs. `IBP_4_0_IV0` is only used for bootstrapping the `group.version` feature flag. The ELR moved to `IBP_4_0_IV1`.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupHeartbeatRequestTest.scala
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupHeartbeatRequestTest.scala:66]: In https://github.com/apache/kafka/pull/16482/files we had a test for the describe request as well. Do we still need this?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:244]: Nice -- this seems a lot simpler/centralized than how we had it before.

**Comment by jolshan** [comment]
There seem to be some test failures. Have we also run the system tests?

**Comment by dajac** [comment]
> There seem to be some test failures. Have we also run the system tests?

I fixed the tests with the last commits. Regarding the system tests, I ran some of the consumer ones locally. Let me start a full run to check them all.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaApis.scala:3816]: there are 3 conditions for consumer group protocol.

1. `groupCoordinator.isNewGroupCoordinator` this will be removed while old group coordinator gets dropped
2. `config.groupCoordinatorRebalanceProtocols.contains(Group.GroupType.CONSUMER)`  this is a static config
3. `groupVersion().isConsumerRebalanceProtocolSupported` this is dynamic (by updating feature)

not sure why we need both condition_2 and condition_3 in controlling the support of consumer group requests. Or `groupCoordinatorRebalanceProtocols` will be removed in the future too?

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:260]: Could you please use `String.join` instead?
```java
                            throw new FormatterException("Unsupported feature: " + featureName +
                                ". Supported features are: " + String.join(", ", nameToSupportedFeature.keySet()));
```

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:250]: there is another PR (#16991) adding similar behavior, but it adds the code to `BootstrapMetadata`. It seems moving the code to `BootstrapMetadata` is easy to write UT (`BootstrapMetadataTests`). WDYT?

**Review by chia7712 — COMMENTED** [review]
@dajac thanks for this patch!

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3816]: The GroupVersion will always increase like the MetadataVersion because we will use it to gate future changes related to groups in the future too (e.g. share groups). So we also have `groupCoordinatorRebalanceProtocols` as a static config to allow admin to disable the new protocol or other future protocols if they really don't want them. Honestly, I don't expect it to be really used but it is nice to have.

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (dajac) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:260]: Sure.

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (dajac) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:250]: Sure. We could have a new static method which create a bootstrap with the default features based on the metadata version. It looks like https://github.com/apache/kafka/pull/16991 does not have the later part if I read it correctly. Another approach would be to keep it as it is and to handle the refactor in https://github.com/apache/kafka/pull/16991 as it is already refactoring it.

Comment (chia7712) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:250]: > Another approach would be to keep it as it is and to handle the refactor in https://github.com/apache/kafka/pull/16991 as it is already refactoring it.

fair enough

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (dajac) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:260]: @chia7712 For my understanding, does it make a difference or is it only a better/preferred syntax?

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (dajac) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:250]: I checked and I confirm that it is easier to do it once https://github.com/apache/kafka/pull/16991 is merged or to do it in https://github.com/apache/kafka/pull/16991. I need some of the code there to implement it.

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (dajac) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:260]: Updated.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (chia7712) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3816]: > So we also have groupCoordinatorRebalanceProtocols as a static config to allow admin to disable the new protocol or other future protocols if they really don't want them. 

However, users must use it if they want to enable `CONSUMER` protocol, since the default value has only `CLASSIC` protocol. Maybe we should revert the config from "enable" to "disable"?

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (chia7712) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:260]: > is it only a better/preferred syntax?

yes :)

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java
Comment (chia7712) [reply core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java:250]: > I checked and I confirm that it is easier to do it once https://github.com/apache/kafka/pull/16991 is merged or to do it in https://github.com/apache/kafka/pull/16991. I need some of the code there to implement it.

ok, let's keep current code. #16991 can refactor code after this PR gets merged.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3816]: Give me a bit of time to finish everything :). My next PR will change the default to `classic,consumer`. The plan is to have the new protocol enabled by default in 4.0. I will do this change separately though because it requires some cleaning too.

**Review by jolshan — APPROVED** [review]
lgtm if tests also look good 👍

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by dajac** [comment]
The unit and integration tests look good. However, I was not able to get a full run of all the system tests. The run always fails with `[Errno 24] Too many open files` error. This also impacts trunk. As I was able to run some of the system tests (consumer and kraft upgrade) locally, I believe that this PR does not break anything. I will merge it and monitor the system tests.

**Comment by jolshan** [comment]
Ugh. I feel like we need to find a better way to keep the system tests healthy :( Thanks for taking a look locally though David.

**Pull request merged**
